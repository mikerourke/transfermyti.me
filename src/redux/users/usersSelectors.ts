import { isNil, propOr } from "ramda";

import { selectIdToLinkedId } from "~/api/selectIdToLinkedId";
import { createSelector } from "~/redux/reduxTools";
import { workspaceIdToLinkedIdSelector } from "~/redux/workspaces/workspacesSelectors";
import type { ReduxState, User } from "~/typeDefs";

const sourceUsersByIdSelector = createSelector(
  (state: ReduxState) => state.users.source,
  (sourceUsersById): Dictionary<User> => sourceUsersById,
);

export const sourceUsersSelector = createSelector(
  sourceUsersByIdSelector,
  (sourceUsersById): User[] => Object.values(sourceUsersById),
);

export const includedSourceUsersSelector = createSelector(
  sourceUsersSelector,
  (sourceUsers): User[] =>
    sourceUsers.filter((sourceUser) => sourceUser.isIncluded),
);

export const userIdToLinkedIdSelector = createSelector(
  sourceUsersByIdSelector,
  (sourceUsersById): Dictionary<string> => selectIdToLinkedId(sourceUsersById),
);

export const sourceUsersForTransferSelector = createSelector(
  includedSourceUsersSelector,
  (sourceUsers): User[] =>
    sourceUsers.filter((sourceUser) => isNil(sourceUser.linkedId)),
);

export const sourceUserEmailsByWorkspaceIdSelector = createSelector(
  sourceUsersSelector,
  workspaceIdToLinkedIdSelector,
  (sourceUsers, workspaceIdToLinkedId): Dictionary<string[]> => {
    const emailsByWorkspaceId: Dictionary<string[]> = {};

    for (const sourceUser of sourceUsers) {
      const targetWorkspaceId = propOr<null, Dictionary<string>, string>(
        null,
        propOr("", "workspaceId", sourceUser) as string,
        workspaceIdToLinkedId,
      );

      if (!isNil(targetWorkspaceId)) {
        const emailsInWorkspace = emailsByWorkspaceId[targetWorkspaceId] ?? [];

        emailsByWorkspaceId[targetWorkspaceId] = [
          ...emailsInWorkspace,
          sourceUser.email,
        ];
      }
    }

    return emailsByWorkspaceId;
  },
);
