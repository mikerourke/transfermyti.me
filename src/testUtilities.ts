export { FAKES } from "~/redux/__fakes__/state";

type MockInstance = jest.MockInstance<AnyValid, AnyValid>;

interface WrapForMockingReturn {
  calls: AnyValid[];
  clear(): MockInstance;
  implementOnce(funcOrReturn: AnyValid): MockInstance;
  rejectAlways(err?: AnyValid): MockInstance;
  rejectOnce(err?: AnyValid): MockInstance;
  resolveMultiple(values: AnyValid[]): void;
  resolveOnce(value?: AnyValid): MockInstance;
  returnOnce(value?: AnyValid): MockInstance;
  returnMultiple(values: AnyValid[]): void;
  throwOnce(message?: string): MockInstance;
}

/**
 * Wrappers around someFunc.mock...() that Jest provides. Using
 * this avoids the annoying `someFunc as AnyValid` required to shut TypeScript
 * up.
 */
export function wrapForMocking(func: AnyValid): WrapForMockingReturn {
  return {
    calls: func?.mock?.calls,
    clear(): MockInstance {
      return func.mockClear();
    },
    implementOnce(funcOrReturn: AnyValid): MockInstance {
      if (typeof funcOrReturn === "function") {
        return func.mockImplementationOnce(funcOrReturn);
      } else {
        return func.mockImplementationOnce(() => funcOrReturn);
      }
    },
    rejectAlways(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValue(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    rejectOnce(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValueOnce(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    resolveMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockResolvedValueOnce(value);
      }
    },
    resolveOnce(value: AnyValid = null): MockInstance {
      return func.mockResolvedValueOnce(value);
    },
    returnOnce(value: AnyValid = null): MockInstance {
      return func.mockReturnValueOnce(value);
    },
    returnMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockReturnValueOnce(value);
      }
    },
    throwOnce(message: string = "error"): MockInstance {
      return func.mockImplementationOnce(() => {
        throw new Error(message);
      });
    },
  };
}

/**
 * Allows you to get around `delay()` calls in sagas.
 * @see https://github.com/jfairbank/redux-saga-test-plan/issues/257#issuecomment-503016758
 * @example
 *   return expectSaga(someSaga)
 *     .provide([
 *       { call: provideDelay },
 *     ])
 *     ...
 */
export function provideDelay(
  { fn }: AnyValid,
  next: AnyValid,
): null | VoidFunction {
  return fn.name === "delayP" ? null : next();
}
