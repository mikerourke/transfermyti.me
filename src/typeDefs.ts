import type { AllEntitiesState } from "~/redux/allEntities/allEntitiesReducer";
import type { AppState } from "~/redux/app/appReducer";
import type { ClientsState } from "~/redux/clients/clientsReducer";
import type { CredentialsState } from "~/redux/credentials/credentialsReducer";
import type { ProjectsState } from "~/redux/projects/projectsReducer";
import type { TagsState } from "~/redux/tags/tagsReducer";
import type { TasksState } from "~/redux/tasks/tasksReducer";
import type { TimeEntriesState } from "~/redux/timeEntries/timeEntriesReducer";
import type { UserGroupsState } from "~/redux/userGroups/userGroupsReducer";
import type { UsersState } from "~/redux/users/usersReducer";
import type { WorkspacesState } from "~/redux/workspaces/workspacesReducer";

export interface ReduxState {
  allEntities: AllEntitiesState;
  app: AppState;
  clients: ClientsState;
  credentials: CredentialsState;
  projects: ProjectsState;
  tags: TagsState;
  tasks: TasksState;
  timeEntries: TimeEntriesState;
  userGroups: UserGroupsState;
  users: UsersState;
  workspaces: WorkspacesState;
}

type NotificationType = "error" | "info" | "success";

export interface Notification {
  id: string;
  message: string;
  type: NotificationType;
}

export enum ToolName {
  None = "none",
  Clockify = "clockify",
  Toggl = "toggl",
}

export enum Mapping {
  Source = "source",
  Target = "target",
}

export enum ToolAction {
  None = "none",
  Delete = "delete",
  Transfer = "transfer",
}

export interface ToolHelpDetails {
  toolName: ToolName;
  displayName: string;
  toolLink: string;
}

export enum EntityGroup {
  Clients = "clients",
  Projects = "projects",
  Tags = "tags",
  Tasks = "tasks",
  TimeEntries = "timeEntries",
  UserGroups = "userGroups",
  Users = "users",
  Workspaces = "workspaces",
}

export enum FetchStatus {
  Pending = "PENDING",
  InProcess = "IN_PROCESS",
  Success = "SUCCESS",
  Error = "ERROR",
}

export interface AnyEntity {
  id: string;
  workspaceId: string;
  linkedId: string | null;
  isIncluded: boolean;
  entryCount: number;
  memberOf?: EntityGroup;
}

export type ValidEntity<TEntity> = TEntity & AnyEntity;

export type EntityTableRecord<TEntity> = TEntity & {
  existsInTarget: boolean;
  isActiveInSource: boolean;
  isActiveInTarget: boolean;
};

export type CountsByEntityGroup = Record<
  Exclude<EntityGroup, EntityGroup.Workspaces>,
  number
>;

export interface ToolNameByMapping {
  [Mapping.Source]: ToolName;
  [Mapping.Target]: ToolName;
}

export interface Client extends AnyEntity {
  id: string;
  name: string;
  workspaceId: string;
}

export type ClientTableRecord = EntityTableRecord<Client> & {
  projectCount: number;
};

export interface Credentials {
  apiKey: string | null;
  email: string | null;
  userId: string | null;
}

export type CredentialsByMapping = Record<Mapping, Credentials>;

export type ValidationErrorsByMapping = Record<Mapping, string | null>;

export interface PartialCredentialsUpdate {
  mapping: Mapping;
  apiKey?: string | null;
  email?: string | null;
  userId?: string | null;
}

export interface Estimate {
  estimate: number | string;
  type: "AUTO" | "MANUAL";
}

export interface Project extends AnyEntity {
  id: string;
  name: string;
  clientId: string | null;
  isBillable: boolean;
  isPublic: boolean;
  isActive: boolean;
  estimate: Estimate;
  color: string;
  userIds: string[];
}

export interface Tag extends AnyEntity {
  id: string;
  name: string;
  workspaceId: string;
}

export interface Task extends AnyEntity {
  id: string;
  name: string;
  estimate: string;
  projectId: string;
  assigneeIds: string[];
  isActive: boolean;
}

export type TaskTableRecord = EntityTableRecord<Task> & {
  projectName: string;
};

export interface TimeEntry extends AnyEntity {
  id: string;
  description: string;
  isBillable: boolean;
  start: Date;
  end: Date;
  year: number;
  isActive: boolean;
  clientId: string | null;
  projectId: string | null;
  tagIds: string[];
  tagNames: string[];
  taskId: string | null;
  userId: string | null;
  userGroupIds: string[];
}

export interface TimeEntryTableRecord extends EntityTableRecord<TimeEntry> {
  taskName: string | null;
  projectName: string | null;
}

export interface UserGroup extends AnyEntity {
  id: string;
  name: string;
  userIds: string[];
}

export interface User extends AnyEntity {
  id: string;
  name: string;
  email: string;
  isAdmin: boolean | null;
  isActive: boolean;
  userGroupIds: string[] | null;
}

export interface Workspace extends AnyEntity {
  id: string;
  name: string;
  userIds: string[];
  isAdmin: boolean | null;
}
