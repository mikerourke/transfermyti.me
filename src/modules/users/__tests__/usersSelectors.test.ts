import cases from "jest-in-case";

import { FAKES } from "~/testUtilities";
import { EntityGroup } from "~/typeDefs";

import * as usersSelectors from "../usersSelectors";

const MOCK_STATE = {
  ...FAKES.REDUX_STATE,
  users: {
    ...FAKES.REDUX_STATE.users,
    source: {
      ...FAKES.REDUX_STATE.users.source,
      "6002": {
        id: "6002",
        name: "Mary Testarossa",
        email: "mary@testarossa.com",
        isAdmin: null,
        isActive: true,
        userGroupIds: [],
        workspaceId: FAKES.TOGGL_WORKSPACE_ID,
        entryCount: 0,
        linkedId: null,
        isIncluded: true,
        memberOf: EntityGroup.Users,
      },
    },
  },
};

describe("within usersSelectors", () => {
  test("the sourceUsersSelector returns array of values from state.source", () => {
    const result = usersSelectors.sourceUsersSelector(MOCK_STATE);

    expect(result).toEqual(Object.values(MOCK_STATE.users.source));
  });

  cases(
    "the selectors match their snapshots",
    (options) => {
      const result = options.selector(MOCK_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceUsersSelector",
        selector: usersSelectors.includedSourceUsersSelector,
      },
      {
        name: "for the userIdToLinkedIdSelector",
        selector: usersSelectors.userIdToLinkedIdSelector,
      },
      {
        name: "for the sourceUsersForTransferSelector",
        selector: usersSelectors.sourceUsersForTransferSelector,
      },
      {
        name: "for the sourceUserEmailsByWorkspaceIdSelector",
        selector: usersSelectors.sourceUserEmailsByWorkspaceIdSelector,
      },
    ],
  );
});
