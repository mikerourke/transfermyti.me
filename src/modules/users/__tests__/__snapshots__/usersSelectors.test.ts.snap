// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`within usersSelectors > the selectors match their snapshots > for the sourceUserEmailsByWorkspaceIdSelector 1`] = `
[
  {
    "email": "mary@testarossa.com",
    "entryCount": 0,
    "id": "6002",
    "isActive": true,
    "isAdmin": null,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "users",
    "name": "Mary Testarossa",
    "userGroupIds": [],
    "workspaceId": "1001",
  },
]
`;

exports[`within usersSelectors > the selectors match their snapshots > for the sourceUserEmailsByWorkspaceIdSelector 2`] = `
{
  "6001": "clock-user-01",
  "clock-user-01": "6001",
}
`;

exports[`within usersSelectors > the selectors match their snapshots > for the sourceUserEmailsByWorkspaceIdSelector 3`] = `
[
  {
    "email": "mary@testarossa.com",
    "entryCount": 0,
    "id": "6002",
    "isActive": true,
    "isAdmin": null,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "users",
    "name": "Mary Testarossa",
    "userGroupIds": [],
    "workspaceId": "1001",
  },
]
`;

exports[`within usersSelectors > the selectors match their snapshots > for the sourceUserEmailsByWorkspaceIdSelector 4`] = `
{
  "clock-workspace-01": [
    "test-user@test.com",
    "mary@testarossa.com",
  ],
}
`;
